compiler/sem/mirtrees.nim

```diff
@@ -0,0 +1,360 @@
+## This module contais the type defitions for the mid-end IR (MIR) together
+## with routines for querying them.
+##
+## Routines for traversing the used flat-tree-structure are also located here.
+
+import
+  compiler/ast/[
+    ast_types
+  ]
+
+type
+  TempId* = distinct uint32 ## ID of a temporary variable
+  BlockId* = distinct int32 # TODO: rename to ``LabelId`` (or similar)
+
+  PathExprNodeKind* = enum
+    # note: each operation has an implicit first parameter
+    penkField ## operation: named field access. Zero parameters
+    penkPos   ## operation: anonymous field access (i.e. indexing into an array
+              ## or tuple). The position is provided by the first argument
+    penkCast  ## operation: zero parameters
+    penkConv  ## operation: zero parameters
+    penkDeref ## XXX: I'm not yet sure if having this one is a good idea
+    penkInt   ## operand: an integer literal
+    penkParam ## operand: references a parameter
+    penkEnd   ## marks the end of a path expression
+
+  PathExprNode* = object
+    ## The building block for path expressions. A path expression is a
+    ## function that takes 1 to n value arguments and produces an lvalue.
+    ##
+    ## The output of each operation is the first argument to the operation
+    ## following. For the first operation in the expression, the functions
+    ## first parameter is used as the input.
+    ## A path expression is pure, that is, the output is only dependent on the
+    ## inputs (arguments)
+    case kind*: PathExprNodeKind
+    of penkField:
+      field*: PSym
+    of penkCast, penkConv:
+      typ*: PType
+    of penkInt:
+      intVal*: BiggestUInt
+    of penkParam:
+      param*: uint8
+    of penkPos, penkDeref, penkEnd:
+      discard
+
+  PathOpId* = distinct uint32
+
+  MirNodeKind* = enum
+    # references to named entities
+    mnkProc
+    mnkConst
+    mnkGlobal
+    mnkParam
+    mnkLocal
+    mnkTemp
+
+    mnkType
+
+    mnkLiteral
+
+    mnkDef     ## may be used as either an input or as a statement. That is, it
+               ## accepts zero or one argument
+
+    mnkReturn, mnkBreak
+
+    mnkFastAsgn
+    mnkInit
+    mnkAsgn
+
+    mnkPathOp ## applies a path operation
+    mnkCall   ## a call to a procedure
+    mnkMagic  ## a call to a magic procedure. Can be used as a sink if it has
+              ## only one operand
+
+    mnkArg    ## argument sink. Only allowed directly inside an ``mnkArgBlock``
+    mnkVoid   ## depending on its usage, can be either a sink (then its used
+              ## to discard its input) or an input yielding the 'void' value
+              ## (used when wanting to provide nothing as an argument to an
+              ## operation)
+    mnkName
+
+    mnkModify ## modify(x)
+    mnkConsume## consume(x);
+              ## - this is the last use of `x`, consumes `x` (what consume
+              ##   means depends on the context) it
+              ## - this is not the last use of `x`: creates a full copy of `x`
```

Comment:
Oh, that doc comment is barely readable - sorry. The `consume` operation takes an unnamed value or l-value as input and produces a value. More specifically:
- if the input is an unnamed value (e.g. the result of a procedure call not returning a view), the output value is the *same* (as in identity) as the input value
- if the input is an l-value and it can be proven that it's not used beyond the point of the `consume` operation, an unnamed value *equal* (i.e. same in-memory representation) of the input l-value is produced and the source is reset (but not destroyed)
- if the input is an l-value that is used beyond the `consume` operation, a full copy of the input l-value is produced

In other words, the value produced by the `consume` operation is always guaranteed to be unique, that is, there doesn't exist another value with the same identity. The `consume` operation was originally named `own` (which I now believe fits better), but for reasons I don't remember, I renamed it.